src/control/security/pem_test.go

```diff
@@ -0,0 +1,161 @@
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package security
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+)
+
+func TestLoadPEMData(t *testing.T) {
+
+	goodCertPath := "appdata/certs/daosCA.crt"
+	badCertPerm := "appdata/certs/badperms.crt"
+	badCertPermError := &insecureError{badCertPerm, SafeCertPerm}
+
+	testCases := []struct {
+		filename string
+		perms    os.FileMode
+		testname string
+		expected string
+	}{
+		{goodCertPath, SafeCertPerm, "GoodCert", ""},
+		{badCertPerm, SafeCertPerm, "ImproperPermissions", badCertPermError.Error()},
+	}
+
+	// Setup permissions for tests below.
+	if err := os.Chmod(goodCertPath, SafeCertPerm); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chmod(badCertPerm, 0666); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.testname, func(t *testing.T) {
+			result := ""
+			_, err := LoadPEMData(tc.filename, tc.perms)
+			if err != nil {
+				result = err.Error()
+			}
+			if strings.Compare(result, tc.expected) != 0 {
+				t.Errorf("result %s; expected %s", result, tc.expected)
+			}
+		})
+	}
+}
+
+func TestLoadPrivateKey(t *testing.T) {
+	goodKeyPath := "appdata/certs/daosCA.key"
+	badKeyPerm := "appdata/certs/badperms.key"
+	malformed := "appdata/certs/bad.key"
+	toomany := "appdata/certs/toomanypem.key"
+	notkey := "appdata/certs/notkey.crt"
+	badKeyPermError := &insecureError{badKeyPerm, SafeKeyPerm}
+	malformedError := fmt.Sprintf("%s does not contain PEM data", malformed)
+	toomanyError := "Only one key allowed per file"
+	notkeyError := "PEM Block is not a Private Key"
+
+	// Setup permissions for tests below.
+	if err := os.Chmod(goodKeyPath, SafeKeyPerm); err != nil {
+		t.Fatal(err)
+	}
+	// Intentionallly safe cert perm as it should be incorrect
+	if err := os.Chmod(badKeyPerm, SafeCertPerm); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chmod(malformed, SafeKeyPerm); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chmod(toomany, SafeKeyPerm); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chmod(notkey, SafeKeyPerm); err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := []struct {
+		filename string
+		testname string
+		expected string
+	}{
+		{goodKeyPath, "GoodKey", ""},
+		{badKeyPerm, "ImproperPermissions", badKeyPermError.Error()},
+		{malformed, "Malformed Key", malformedError},
+		{toomany, "InvalidPEMCount", toomanyError},
+		{notkey, "NotAKey", notkeyError},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.testname, func(t *testing.T) {
+			result := ""
+			_, err := LoadPrivateKey(tc.filename)
+			if err != nil {
+				result = err.Error()
+			}
+			if strings.Compare(result, tc.expected) != 0 {
```

Comment:
Any reason not to use the AssertEqual function instead of strings.Compare?

Reply:
It's because an AssertEqual wouldn't allow the rest of the subcases to run. According to the documentation it would exit upon failing the Assertion.

Reply:
We have used Assert* in other unit tests, and the test suite halts with an error upon a single assert, rather than continuing to other tests.  I thought that was intentional.  Can you check with Tom or Michael to see what the correct behavior ought to be?

Reply:
Since we're using table based tests I don't think Assert is the right thing. Unless that failure will cause all other tests to fail as well Assert in that case is heavy handed and not helpful in terms of running a regression test suite since you can get two test runs that are incomparable.